src/network/token_provider.rs

```diff
@@ -64,14 +68,76 @@ impl TokenProvider for IBMCloudTokenProvider {
         form_data.insert("apikey".to_string(), self.apikey.to_string());
 
         let client = Client::new();
-        Ok(client
+        let new_token = client
             .post("https://iam.cloud.ibm.com/identity/token")
             .header("Accept", "application/json")
             .form(&form_data)
             .send()
             .map_err(NetworkError::ReqwestError)?
             .json::<AccessTokenResponse>()
-            .map_err(NetworkError::ReqwestError)? // FIXME: This is a deserialization error (extract it from Reqwest)
-            .access_token)
+            .map_err(NetworkError::ReqwestError)?; // FIXME: This is a deserialization error (extract it from Reqwest)
+
+        let mut access_token = self.access_token.write()?;
+        access_token.renew(new_token.access_token, new_token.expires_in);
+        Ok(())
+    }
+}
+
+#[derive(Deserialize)]
+struct AccessTokenResponse {
+    access_token: String,
+    expires_in: u64,
+    // We are discarding some fields:
+    // refresh_token: String, // "not_supported"
+    // token_type: String,    // "Bearer"
+    // expiration: u64,       // 1755854106 <- unix time when it expires
+    // scope: String,         // "ibm openid"
+}
+
+impl TokenProvider for IBMCloudTokenProvider {
+    fn get_access_token(&self) -> NetworkResult<String> {
+        if self.access_token.read()?.expired() {
+            self.renew_token()?;
+        }
+
+        Ok(self.access_token.read()?.token.to_string())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_access_token() {
+        let mut access_token = AccessToken::default();
+        assert!(access_token.expired());
+
+        access_token.renew("something".to_string(), 10);
+        assert!(!access_token.expired());
+        assert_eq!(access_token.token, "something".to_string());
```

Comment:
should we add more tests around the timings? to check that current time is honored and the expire time is honored?